src/ir/type-updating.h

```diff
@@ -434,13 +439,26 @@ class TypeMapper : public GlobalTypeRewriter {
     : GlobalTypeRewriter(wasm), mapping(mapping) {}
 
   void map() {
-    // Map the types of expressions (curr->type, etc.) to their merged
-    // types.
-    mapTypes(mapping);
-
     // Update the internals of types (struct fields, signatures, etc.) to
     // refer to the merged types.
-    update();
+    auto oldToNewTypes = rebuildTypes();
+
+    // Compose the user-provided mapping from old types to other old types with
+    // the new mapping from old types to new types. `oldToNewtypes` will become
+    // a copy of `mapping` except that the destination types will be the newly
+    // built types.
+    for (auto& [src, dest] : mapping) {
+      if (auto it = oldToNewTypes.find(dest); it != oldToNewTypes.end()) {
+        oldToNewTypes[src] = it->second;
```

Comment:
I'm a little confused here. `mapping` is the user provided mapping of types existing before this operation, and `oldToNewTypes` maps old types to rebuilt types. The operation below can make `oldToNewTypes[..]` be equal to a type existing before this operation (since it copies `dest` from `mapping`), which sounds wrong? I'd expect something like
```suggestion
        oldToNewTypes[src] = oldToNewTypes[it->second];
```
That is, `mapping` tells us what original type we want, but we need to apply `oldToNewTypes` on the right in order to get the the rebuilt version of that type. Is that wrong?

Side note, could `oldToNewTypes` here be renamed to `rebuiltTypes` or such? (old/new has more than one meaning here.)

Reply:
Yeah, renaming `oldToNewTypes` to `newMapping` might make this even clearer, since it ends up taking the place of the original `mapping`.

Reply:
Regarding the code you would expect, since `it` is result of looking up `dest` in `oldToNewTypes`, we're already doing the right-hand-side `oldToNewTypes` lookup you are expecting to see. Doing an additional lookup of `oldToNewTypes[it->second]` wouldn't make sense because `it->second` is already a newly built type.

Reply:
Ah, thanks, I was indeed confused here.